src/components/ModalSignup.tsx

```diff
@@ -0,0 +1,177 @@
+import React, { ChangeEvent, FC, useCallback, useEffect, useState } from 'react'
+import { connect, useDispatch } from 'react-redux'
+import { ActionButton } from './ActionButton'
+import { Index, Connected, State } from '../@types'
+import { showModal } from '../action-creators'
+import Modal from './Modal'
+import { storage } from '../util/storage'
+
+const firebaseErrorsIndex = {
+  'signup/no-code':
+    'em is currently in beta. Please join the waitlist, or find a friendly person who can offer you an invite to gain access immediately.',
+  'signup/invalid-code': 'Invalid invitation code',
+  'signup/already-used-code': 'This invitation code has already been used',
+  default: 'Something went wrong',
+}
+
+type errorCode = keyof typeof firebaseErrorsIndex
+
+type SubmitAction = (closeModal: () => void, email: string, password?: string) => Promise<void>
+
+interface InputProps {
+  type: 'email' | 'password'
+  placeholder: string
+  onBlur: (e: ChangeEvent<HTMLInputElement>) => void
+  value: string
+}
+/**
+ *
+ */
+const Input: FC<InputProps> = ({ type, placeholder, value, onBlur }) => {
+  const [inputValue, updateInputValue] = useState(value)
+
+  useEffect(() => {
+    updateInputValue(value)
+  }, [value])
+
+  /** On input change handler. */
+  const onChange = (e: ChangeEvent<HTMLInputElement>) => updateInputValue(e.target.value)
+
+  return <input type={type} placeholder={placeholder} value={inputValue} onChange={onChange} onBlur={onBlur} />
+}
+
+interface Mode {
+  name: string
+  modalKey: string
+  modalTitle: string
+}
+
+const modes: Index<Mode> = {
+  signup: {
+    name: 'signup',
+    modalKey: 'signup',
+    modalTitle: 'Sign Up',
+  },
+}
+
+// eslint-disable-next-line jsdoc/require-jsdoc
+const mapStateToProps = (state: State) => {
+  const { invitationCode, invitationCodeDetail } = state
+  return {
+    invitationCode,
+    invitationCodeDetail,
+  }
+}
+
+/** A modal dialog for signing up. */
+const ModalSignup = ({ invitationCode, invitationCodeDetail }: Connected<ReturnType<typeof mapStateToProps>>) => {
+  const dispatch = useDispatch()
+  const [email, updateEmail] = useState('')
+  const [password, updatePassword] = useState('')
+  const [error, updateError] = useState<null | string>(null)
+
+  const [isSubmitting, updateIsSubmitting] = useState(false)
+
+  useEffect(() => {
+    const hasInvite = invitationCodeDetail && Object.keys(invitationCodeDetail).length === 0
+    let errorCode = ''
+    if (invitationCode === '') {
+      errorCode = 'signup/no-code'
+    } else if (hasInvite) {
+      errorCode = 'signup/invalid-code'
+    } else if (invitationCodeDetail && invitationCodeDetail.usedBy) {
+      errorCode = 'signup/already-used-code'
+    }
```

Comment:
Prefer using the `ternary` operator to replace `if` statements whenever possible.